transport/internal/tlsmux/tls_checker.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tlsmux
+
+import "io"
+
+const (
+	_tlsHandshakeHeaderLength = 5
+
+	_tlsContentTypeHandshake = 22
+	_tlsMajorVersion         = 3
+	_tlsMinorVersion         = 1
```

Comment:
is this the wrong order unless it is for SSL version? We are using TLS 1.3 so Major is 1 and Minor is 3?

Reply:
I think the handshake reflects versioning based on the prior SSL standard.
TLS 1.0 is a rename of SSL 3.1.
Very confusing indeed.

https://tls13.ulfheim.net/:
> 03 01 - protocol version is "3,1" (also known as TLS 1.0)

If the goal is to support only TLS 1.3, then thev alue in the first five bytes should be `0x0303` per https://datatracker.ietf.org/doc/html/rfc8446#section-4.1.2:
> In TLS 1.3, ... the
> legacy_version field MUST be set to 0x0303

If the goal is to support multiple TLS versions, then do not check the minor version at offset 2.

Reply:
Above `isTLSClientHelloRecord` you link section 5.1 of RFC 8446 which is even more detailed about the byte values:

> legacy_record_version:  MUST be set to 0x0303 for all records
> generated by a TLS 1.3 implementation other than an initial
> ClientHello (i.e., one not generated after a HelloRetryRequest),
> where it MAY also be 0x0301 for compatibility purposes.  This
> field is deprecated and MUST be ignored for all purposes.
> Previous versions of TLS would use other values in this field
> under some circumstances.

What is the benefit of checking major and minor version here? For the purpose of "is this TLS" can it be sufficient to check the content type is `22` and leave version negotiation to elsewhere in the stack?

Reply:
Thanks, Josh. TLS versions starting with TLS1.0 are 0x0301, and TLS1.3 is 0x0304. The major version in the TLS header is always 3, and the minor version must always be >= 1. 

> If the goal is to support multiple TLS versions, then do not check the minor version at offset 2.
>What is the benefit of checking major and minor version here? For the purpose of "is this TLS" can it be sufficient to check the content type is 22 and leave version negotiation to elsewhere in the stack?'

Technically, yes. The first byte == 22 is a good signal, but I prefer sticking to the spec to confirm major and minor versions. And yes, negotiation is left to the TLS handler offered by the standard go package.